File: backend/app/schemas/asset.py
from enum import StrEnum
from datetime import date
from decimal import Decimal

from typing import Literal

from pydantic import BaseModel, Field


class AssetSnapshotTotalsResponse(BaseModel):
    snapshot_date: date
    asset_total: Decimal
    liability_total: Decimal
    net_worth: Decimal


class NetWorthBreakdownItemResponse(BaseModel):
    side: str
    category: str
    amount: Decimal
    ratio: float | None


class NetWorthBreakdownResponse(BaseModel):
    snapshot_date: date | None
    asset_total: Decimal
    negative_asset_excluded_total: Decimal = Decimal("0")
    liability_total: Decimal
    net_worth: Decimal
    items: list[NetWorthBreakdownItemResponse]


class AssetLiabilityHealthResponse(BaseModel):
    snapshot_date: date | None
    cash_equivalent_total: Decimal
    asset_total: Decimal
    negative_asset_excluded_total: Decimal = Decimal("0")
    liability_total: Decimal
    net_worth: Decimal
    monthly_required_spend: Decimal
    monthly_required_spend_source: str = "manual_query_param"
    emergency_fund_months: float | None
    emergency_fund_target_months: int
    target_progress_ratio: float | None
    monthly_debt_payment: Decimal
    monthly_income: Decimal
    monthly_income_source: str = "manual_query_param"
    derived_from_periods: list[str] = Field(default_factory=list)
    manual_input_overrides: list[str] = Field(default_factory=list)
    debt_payment_ratio: float | None
    debt_to_asset_ratio: float | None
    confidence: str
    assumptions: list[str]


class AssetLiquidityPatchRequest(BaseModel):
    liquidity_tier: Literal["immediate", "near_liquid", "illiquid"] | None = None
    is_cash_equivalent: bool | None = None


class AssetSnapshotItemResponse(BaseModel):
    id: int
    snapshot_date: date
    side: str
    category: str
    product_name: str
    amount: Decimal
    liquidity_tier: str | None
    is_cash_equivalent: bool | None


class AssetSnapshotsResponse(BaseModel):
    items: list[AssetSnapshotTotalsResponse]
    asset_items: list[AssetSnapshotItemResponse] = Field(default_factory=list)


class SnapshotComparisonMode(StrEnum):
    LATEST_AVAILABLE_VS_PREVIOUS_AVAILABLE = "latest_available_vs_previous_available"
    LAST_CLOSED_MONTH_VS_PREVIOUS_CLOSED_MONTH = (
        "last_closed_month_vs_previous_closed_month"
    )
    SELECTED_SNAPSHOT_VS_BASELINE_SNAPSHOT = "selected_snapshot_vs_baseline_snapshot"


class AssetSnapshotComparisonDeltaResponse(BaseModel):
    asset_total: Decimal
    liability_total: Decimal
    net_worth: Decimal
    asset_total_pct: float | None
    liability_total_pct: float | None
    net_worth_pct: float | None


class AssetSnapshotComparisonResponse(BaseModel):
    comparison_mode: SnapshotComparisonMode
    current: AssetSnapshotTotalsResponse | None
    baseline: AssetSnapshotTotalsResponse | None
    delta: AssetSnapshotComparisonDeltaResponse | None
    comparison_days: int | None
    is_partial: bool
    is_stale: bool
    can_compare: bool
    comparison_label: str


class NetWorthPointResponse(BaseModel):
    snapshot_date: date
    net_worth: Decimal


class NetWorthHistoryResponse(BaseModel):
    items: list[NetWorthPointResponse]


class InvestmentItemResponse(BaseModel):
    product_type: str | None
    broker: str
    product_name: str
    cost_basis: Decimal | None
    market_value: Decimal | None
    return_rate: Decimal | None
    pct_of_investment_total: float | None = None


class InvestmentTotalsResponse(BaseModel):
    cost_basis: Decimal
    market_value: Decimal


class InvestmentSummaryResponse(BaseModel):
    snapshot_date: date | None
    items: list[InvestmentItemResponse]
    totals: InvestmentTotalsResponse


class InsuranceContractItemResponse(BaseModel):
    id: int
    snapshot_date: date
    insurer: str
    product_name: str
    contract_status: str | None
    total_paid: Decimal | None
    contract_date: date | None
    maturity_date: date | None


class InsurancePremiumEstimateResponse(BaseModel):
    period: str | None
    amount: Decimal | None
    assumptions: list[str]
    basis: dict[str, object] | None = None


class InsuranceSummaryResponse(BaseModel):
    snapshot_date: date | None
    has_contract_snapshot: bool = False
    missing_reason: str | None = None
    expected_source: str = "BankSalad 4.보험현황"
    items: list[InsuranceContractItemResponse]
    monthly_premium_estimate: InsurancePremiumEstimateResponse


class LoanItemResponse(BaseModel):
    id: int | None = None
    loan_type: str | None
    lender: str
    product_name: str
    principal: Decimal | None
    balance: Decimal | None
    interest_rate: Decimal | None
    monthly_payment: Decimal | None = None
    repayment_method: str | None = None
    monthly_payment_source: str | None = None
    repayment_method_source: str | None = None
    loan_kind: str | None = None
    start_date: date | None
    maturity_date: date | None


class LoanRepaymentMetadataPatchRequest(BaseModel):
    monthly_payment: Decimal | None = Field(default=None, ge=0)
    repayment_method: (
        Literal[
            "principal_interest",
            "principal_equal",
            "interest_only",
            "unknown",
        ]
        | None
    ) = None


class LoanRepaymentMetadataResponse(BaseModel):
    id: int
    snapshot_date: date
    lender: str
    product_name: str
    monthly_payment: Decimal | None
    repayment_method: str | None
    monthly_payment_source: str | None = None
    repayment_method_source: str | None = None


class LoanTotalsResponse(BaseModel):
    principal: Decimal
    balance: Decimal


class LoanSummaryResponse(BaseModel):
    snapshot_date: date | None
    as_of_date: date | None = None
    summary_scope: str = "active_loans_only"
    excluded_historical_count: int = 0
    items: list[LoanItemResponse]
    totals: LoanTotalsResponse

File: frontend/src/types/asset.ts
import type { LoanKind } from './transaction'
export type { LoanKind } from './transaction'

export interface AssetSnapshotTotals {
  id?: number
  snapshot_date: string
  side?: 'asset' | 'liability' | string
  category?: string | null
  product_name?: string | null
  amount?: string | null
  asset_total: string    // Decimal as string
  liability_total: string
  net_worth: string
  liquidity_tier?: LiquidityTier | null
  is_cash_equivalent?: boolean | null
}

export interface AssetSnapshotsResponse {
  items: AssetSnapshotTotals[]
  asset_items: AssetSnapshotItemResponse[]
}

export type LiquidityTier = 'immediate' | 'near_liquid' | 'illiquid'

export interface AssetLiquidityPatchRequest {
  liquidity_tier: LiquidityTier | null
  is_cash_equivalent: boolean | null
}

export interface AssetSnapshotItemResponse {
  id: number
  snapshot_date: string
  side: string
  category: string
  product_name: string
  amount: string
  liquidity_tier: LiquidityTier | null
  is_cash_equivalent: boolean | null
}

export type SnapshotComparisonMode =
  | 'latest_available_vs_previous_available'
  | 'last_closed_month_vs_previous_closed_month'
  | 'selected_snapshot_vs_baseline_snapshot'

export interface AssetSnapshotComparisonDelta {
  asset_total: string
  liability_total: string
  net_worth: string
  asset_total_pct: number | null
  liability_total_pct: number | null
  net_worth_pct: number | null
}

export interface AssetSnapshotComparisonResponse {
  comparison_mode: SnapshotComparisonMode
  current: AssetSnapshotTotals | null
  baseline: AssetSnapshotTotals | null
  delta: AssetSnapshotComparisonDelta | null
  comparison_days: number | null
  is_partial: boolean
  is_stale: boolean
  can_compare: boolean
  comparison_label: string
}

export interface NetWorthPoint {
  snapshot_date: string
  net_worth: string
}

export interface NetWorthHistoryResponse {
  items: NetWorthPoint[]
}

export interface InvestmentItem {
  product_type: string | null
  broker: string
  product_name: string
  cost_basis: string | null
  market_value: string | null
  return_rate: string | null
  pct_of_investment_total: number | null
}

export interface InvestmentSummaryResponse {
  snapshot_date: string | null
  items: InvestmentItem[]
  totals: { cost_basis: string; market_value: string }
}

export interface InsuranceContractItem {
  id: number
  snapshot_date: string
  insurer: string
  product_name: string
  contract_status: string | null
  total_paid: string | null
  contract_date: string | null
  maturity_date: string | null
}

export interface InsurancePremiumEstimate {
  period: string | null
  amount: string | null
  assumptions: string[]
  basis?: Record<string, unknown> | null
}

export interface InsuranceSummaryResponse {
  snapshot_date: string | null
  has_contract_snapshot: boolean
  missing_reason: string | null
  expected_source: string
  items: InsuranceContractItem[]
  monthly_premium_estimate: InsurancePremiumEstimate
}

export interface LoanItem {
  id?: number | null
  loan_type: string | null
  lender: string
  product_name: string
  principal: string | null
  balance: string | null
  interest_rate: string | null
  monthly_payment?: string | null
  repayment_method?: LoanRepaymentMethod | null
  monthly_payment_source?: LoanRepaymentMetadataSource | null
  repayment_method_source?: LoanRepaymentMetadataSource | null
  loan_kind?: LoanKind | null
  start_date: string | null
  maturity_date: string | null
}

export type LoanRepaymentMetadataSource =
  | 'manual'
  | 'estimated_from_linked_transactions'
  | 'derived_from_loan_account'

export type LoanRepaymentMethod =
  | 'principal_interest'
  | 'principal_equal'
  | 'interest_only'
  | 'unknown'

export interface LoanRepaymentMetadataPatchRequest {
  monthly_payment?: string | null
  repayment_method?: LoanRepaymentMethod | null
}

export interface LoanRepaymentMetadataResponse {
  id: number
  snapshot_date: string
  lender: string
  product_name: string
  monthly_payment: string | null
  repayment_method: LoanRepaymentMethod | null
  monthly_payment_source: LoanRepaymentMetadataSource | null
  repayment_method_source: LoanRepaymentMetadataSource | null
}

export interface LoanSummaryResponse {
  snapshot_date: string | null
  as_of_date?: string | null
  summary_scope?: string
  excluded_historical_count?: number
  items: LoanItem[]
  totals: { principal: string; balance: string }
}

export interface NetWorthBreakdownItem {
  side: string
  category: string
  amount: string
  ratio: number | null
}

export interface NetWorthBreakdownResponse {
  snapshot_date: string | null
  asset_total: string
  negative_asset_excluded_total: string
  liability_total: string
  net_worth: string
  items: NetWorthBreakdownItem[]
}

export interface AssetLiabilityHealthResponse {
  snapshot_date: string | null
  cash_equivalent_total: string
  asset_total: string
  negative_asset_excluded_total: string
  liability_total: string
  net_worth: string
  monthly_required_spend: string
  monthly_required_spend_source: string
  emergency_fund_months: number | null
  emergency_fund_target_months: number
  target_progress_ratio: number | null
  monthly_debt_payment: string
  monthly_income: string
  monthly_income_source: string
  derived_from_periods: string[]
  manual_input_overrides: string[]
  debt_payment_ratio: number | null
  debt_to_asset_ratio: number | null
  confidence: string
  assumptions: string[]
}
